eventlog/eventlog.py

```diff
@@ -0,0 +1,759 @@
+    def __init__(self, algid: int, buffer: bytes, idx: int):
+        self.algid = Digest(algid)
+        assert self.algid in EfiEventDigest.hashalgmap
+        self.hashalg     = EfiEventDigest.hashalgmap[self.algid]()
+        self.digest_size = self.hashalg.digest_size
+        self.digest      = buffer[idx:idx+self.digest_size]
+
+    # JSON converter -- returns something that can be encoded as JSON
+    def toJson(self):
+        return { 'AlgorithmId': self.algid.name, 'Digest': self.digest.hex() }
+
+    # ----------------------------------------
+    # parse a list of digests in the event log
+    # ----------------------------------------
+    # inputs:
+    #   digestcount: how many digests to parse
+    #   idx: index in the buffer we are parsing
+    #   buffer: input buffer we are parsing
+    # outputs:
+    #   idx: index of first unparsed byte in buffer
+    #   digests: list of parsed digests
+
+    @classmethod
+    def parselist (cls, digestcount:int, buffer: bytes, idx: int) -> Tuple[dict, int]:
+        digests = {}
+        for _ in range(0,digestcount):
+            (algid,)=struct.unpack('<H',buffer[idx:idx+2])
+            digest = EfiEventDigest(algid, buffer, idx+2)
+            digests[algid] = digest
+            idx += 2 + digest.digest_size
+        return digests, idx
+
+# ########################################
+# base class for all EFI events.
+# ########################################
+# GenericEvent is the superclass of all Events
+# parsed by this code. A parsed Event ends up
+# being a GenericEvent type if there is no specialized
+# parser to further interpret it.
+# ########################################
+# * Each Event type has a constructor, which parses the input
+#   buffer. The constructor is used when there is no doubt about
+#   what should be parsed and how it should be interpreted.
+# * However, the main parser entry point is the Parse class method.
+#   This method peeks into the input buffer and makes decisions
+#   about what object the buffer should be parsed into, then invokes
+#   the appropriate constructor for that type.
+#   Examples of decision points include checking EFI variable names,
+#   e.g. "BootOrder" is handled differently from "Boot0001".
+# * Each Event class has a "validate" method used in
+#   testing its internal consistency (not really available for all
+#   event types, so in those cases testing returns "true")
+# * Each Event class has a "toJson" method to convert it into
+#   data structures accepted by the JSON pickler
+#   (i.e. dictionaries, lists, strings)
+
+# NOTE on JSON output of raw events. tpm2_eventlog limits raw event
+# output to 1024 characters when there is no intelligent processing
+# done on it.  We follow suit for reasons of compatibility -- and
+# because no one in their right mind will process a 6kB raw string in
+# JSON/hex.
+
+class GenericEvent:
+    def __init__ (self, eventheader: Tuple[int, int, dict, int, int], buffer: bytes, idx: int):
```

Comment:
IMHO do not use a tuple here, use a dataclass or different named parameters

Reply:
I have not acted on this yet.